packages/blocks/src/api/parser/test/index.js

```diff
@@ -0,0 +1,351 @@
+				attrs: { fruit: 'Bananas' },
+			} );
+			expect( block.name ).toEqual( 'core/test-block' );
+			expect( block.attributes ).toEqual( { fruit: 'Bananas' } );
+		} );
+
+		it( 'should create the requested block with no attributes if it exists', () => {
+			registerBlockType( 'core/test-block', defaultBlockSettings );
+
+			const block = parseRawBlock( {
+				blockName: 'core/test-block',
+				innerHTML: '',
+			} );
+			expect( block.name ).toEqual( 'core/test-block' );
+			expect( block.attributes ).toEqual( {} );
+		} );
+
+		it( 'should fall back to the unregistered type handler for unregistered blocks if present', () => {
+			registerBlockType(
+				'core/unregistered-block',
+				unknownBlockSettings
+			);
+			setUnregisteredTypeHandlerName( 'core/unregistered-block' );
+
+			const block = parseRawBlock( {
+				blockName: 'core/test-block',
+				innerHTML: 'Bananas',
+				attrs: { fruit: 'Bananas' },
+			} );
+			expect( block.name ).toBe( 'core/unregistered-block' );
+			expect( block.attributes.content ).toContain( 'wp:test-block' );
+		} );
+
+		it( 'should fall back to the freeform content handler if block type not specified', () => {
+			registerBlockType( 'core/freeform-block', unknownBlockSettings );
+			setFreeformContentHandlerName( 'core/freeform-block' );
+
+			const block = parseRawBlock( {
+				innerHTML: 'content',
+			} );
+			expect( block.name ).toEqual( 'core/freeform-block' );
+			expect( block.attributes ).toEqual( { content: '<p>content</p>' } );
+		} );
+
+		it( 'should not create a block if no unknown type handler', () => {
+			const block = parseRawBlock( {
+				blockName: 'core/test-block',
+				innerHTML: '',
+			} );
+			expect( block ).toBeUndefined();
+		} );
+
+		it( 'should fallback to an older version of the block if the current one is invalid', () => {
+			registerBlockType( 'core/test-block', {
+				...defaultBlockSettings,
+				attributes: {
+					fruit: {
+						type: 'string',
+						source: 'text',
+						selector: 'div',
+					},
+				},
+				save: ( { attributes } ) => <div>{ attributes.fruit }</div>,
+				deprecated: [
+					{
+						attributes: {
+							fruit: {
+								type: 'string',
+								source: 'text',
+								selector: 'span',
+							},
+						},
+						save: ( { attributes } ) => (
+							<span>{ attributes.fruit }</span>
+						),
+						migrate: ( attributes ) => ( {
+							fruit: 'Big ' + attributes.fruit,
+						} ),
+					},
+				],
+			} );
+
+			const block = parseRawBlock( {
+				blockName: 'core/test-block',
+				innerHTML: '<span>Bananas</span>',
+				attrs: { fruit: 'Bananas' },
+			} );
+			expect( block.name ).toEqual( 'core/test-block' );
+			expect( block.attributes ).toEqual( { fruit: 'Big Bananas' } );
+			expect( block.isValid ).toBe( true );
+			expect( console ).toHaveInformed();
+		} );
+	} );
+
+	describe( 'parse() of @wordpress/block-serialization-spec-parser', () => {
+		// run the test cases using the PegJS defined parser
+		testCases( parsePegjs );
```

Comment:
While we are at it, we can now inline `testCases`. `parseWithGrammar` is no longer there and we also don't test the other parser anymore.